src/descriptor.c

```diff
@@ -0,0 +1,2323 @@
+}
+
+static uint64_t poly_mod_descriptor_checksum(uint64_t c, int val)
+{
+    uint8_t c0 = c >> 35;
+    c = ((c & 0x7ffffffff) << 5) ^ val;
+    if (c0 & 1) c ^= 0xf5dee51989;
+    if (c0 & 2) c ^= 0xa9fdca3312;
+    if (c0 & 4) c ^= 0x1bab10e32d;
+    if (c0 & 8) c ^= 0x3706b1677a;
+    if (c0 & 16) c ^= 0x644d626ffd;
+    return c;
+}
+
+static int generate_checksum(const char *str, size_t str_len, char *checksum_out)
+{
+    uint64_t c = 1;
+    int cls = 0;
+    int clscount = 0;
+    size_t pos;
+    size_t i;
+
+    for (i = 0; i < str_len; ++i) {
+        if ((pos = checksum_get_position(str[i])) == 0)
+            return WALLY_EINVAL; /* Invalid character */
+        --pos;
+        /* Emit a symbol for the position inside the group, for every character. */
+        c = poly_mod_descriptor_checksum(c, pos & 31);
+        /* Accumulate the group numbers */
+        cls = cls * 3 + (int)(pos >> 5);
+        if (++clscount == 3) {
+            /* Emit an extra symbol representing the group numbers, for every 3 characters. */
+            c = poly_mod_descriptor_checksum(c, cls);
+            cls = 0;
+            clscount = 0;
+        }
+    }
+    if (clscount > 0)
+        c = poly_mod_descriptor_checksum(c, cls);
+    for (i = 0; i < DESCRIPTOR_CHECKSUM_LENGTH; ++i)
+        c = poly_mod_descriptor_checksum(c, 0);
+    c ^= 1;
+
+    for (i = 0; i < DESCRIPTOR_CHECKSUM_LENGTH; ++i)
+        checksum_out[i] = checksum_charset[(c >> (5 * (7 - i))) & 31];
+    checksum_out[DESCRIPTOR_CHECKSUM_LENGTH] = '\0';
+
+    return WALLY_OK;
+}
+
+static inline bool is_identifer_char(char c)
+{
+    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_';
+}
+
+static const struct wally_map_item *lookup_identifier(const struct wally_map *map_in,
+                                                      const char *key, size_t key_len)
+{
+    size_t i;
+    for (i = 0; i < map_in->num_items; ++i) {
+        const struct wally_map_item *item = &map_in->items[i];
+        if (key_len == item->key_len - 1 && memcmp(key, item->key, key_len) == 0)
+            return item;
+    }
+    return NULL;
+}
+
+int wally_descriptor_canonicalize(const char *descriptor,
+                                  const struct wally_map *vars_in, uint32_t flags,
+                                  char **output)
+{
+    const size_t VAR_MAX_NAME_LEN = 16;
+    size_t required_len = 0;
+    const char *p = descriptor, *start;
+    char *out;
+
+    if (output)
+        *output = NULL;
+
+    if (!descriptor || flags || !output)
+        return WALLY_EINVAL;
+
+    /* First, find the length of the canonicalized descriptor */
+    while (*p && *p != '#') {
+        while (*p && *p != '#' && !is_identifer_char(*p)) {
+            ++required_len;
+            ++p;
+        }
+        start = p;
+        while (is_identifer_char(*p))
+            ++p;
+        if (p != start) {
+            const bool is_number = *start >= '0' && *start <= '9';
```

Comment:
It seems that only the beginning of the buffer is being checked. I think you should check the whole range of lookup_len.

Reply:
@k-matsuzawa This check just looks at the first digit because variable identifiers must start with a letter or underscore; if the string of characters starts with a number (or if its too long to be an identifier) we avoid looking it up for variable substitution. So for example, `key_1` is a valid variable name, but `1_key` is not (variables use the same naming constraints as e.g. python, c and c++, as well as additionally being limited to 16 characters in length).

I'll rename the variable to `starts_with_digit` and update the comment to make this clearer, thanks!